File: toy_ray_tracer/src/scenes/scene1.rs
use std::sync::Arc;

use crate::camera::{Camera, CameraOpt};
use crate::hittable::Hittable;
use crate::hittable_list::HittableList;
use crate::materials::{Dielectric, Lambertian, Metal};
use crate::scene::{RenderOptions, Scene};
use crate::shapes::{MovingSphere, Sphere, BVH};
use crate::textures::{CheckerTexture, ConstantTexture};
use crate::utils::random;
use crate::vec::{vec3, Vec3};

fn create_world() -> Box<dyn Hittable> {
    let origin = Vec3::new(4.0, 0.2, 0.0);
    let mut world = HittableList::new();
    let checker = Arc::new(CheckerTexture::new(
        Arc::new(ConstantTexture::new(0.2, 0.3, 0.1)),
        Arc::new(ConstantTexture::new(0.9, 0.9, 0.9)),
    ));
    world.add(Sphere::new(
        Vec3::new(0.0, -1000.0, 0.0),
        1000.0,
        Arc::new(Lambertian::new(checker)),
    ));
    for a in -10..10 {
        for b in -10..10 {
            let choose_material = random::f32();
            let center = Vec3::new(
                a as f32 + 0.9 * random::f32(),
                0.2,
                b as f32 + 0.9 * random::f32(),
            );
            if (center - origin).magnitude() > 0.9 {
                if choose_material < 0.8 {
                    // diffuse
                    world.add(MovingSphere::new(
                        center,
                        center + Vec3::new(0.0, 0.5 * random::f32(), 0.0),
                        0.0,
                        1.0,
                        0.2,
                        Arc::new(Lambertian::new(Arc::new(ConstantTexture::new(
                            random::f32() * random::f32(),
                            random::f32() * random::f32(),
                            random::f32() * random::f32(),
                        )))),
                    ));
                } else if choose_material < 0.95 {
                    // metal
                    world.add(Sphere::new(
                        center,
                        0.2,
                        Arc::new(Metal::new(
                            Vec3::new(
                                0.5 * (1.0 + random::f32()),
                                0.5 * (1.0 + random::f32()),
                                0.5 * (1.0 + random::f32()),
                            ),
                            0.5 * random::f32(),
                        )),
                    ));
                } else {
                    // glass
                    world.add(Sphere::new(center, 0.2, Arc::new(Dielectric::new(1.5))));
                }
            }
        }
    }
    world.add(Sphere::new(
        Vec3::new(0.0, 1.0, 0.0),
        1.0,
        Arc::new(Dielectric::new(1.5)),
    ));
    world.add(Sphere::new(
        Vec3::new(-4.0, 1.0, 0.0),
        1.0,
        Arc::new(Lambertian::new(Arc::new(ConstantTexture::new(
            0.4, 0.2, 0.1,
        )))),
    ));
    world.add(Sphere::new(
        Vec3::new(4.0, 1.0, 0.0),
        1.0,
        Arc::new(Metal::new(Vec3::new(0.7, 0.6, 0.5), 0.0)),
    ));
    Box::new(BVH::new(world.move_list(), 0.0, 1.0))
}

pub fn create_scene(opt: RenderOptions) -> Scene {
    let camera = Camera::new(CameraOpt {
        look_from: Vec3::new(13.0, 2.0, 3.0),
        look_at: Vec3::zeros(),
        view_up: vec3::YUP,
        vertical_fov: 20.0,
        aspect: opt.aspect(),
        aperture: 0.0,
        focus_dist: 10.0,
        time0: 0.0,
        time1: 1.0,
    });

    let world = create_world();

    return Scene::new(
        camera,
        world,
        Vec3::new(0.7, 0.8, 1.0),
        String::from("scene1"),
        String::from("random sphere"),
    );
}

File: assets/projects/scene1.js
/// <reference path="../schemas/global.d.ts" />

/** @typedef {import("../schemas/project").Vec3F} Vec3F */

const scenes = [].map((v) => make_scene(v));

function random_float(min = 0, max = 1) {
  return min + Math.random() * (max - min);
}

/** @returns {Vec3F} */
function random_vec3(min = 0, max = 1) {
  return [
    random_float(min, max),
    random_float(min, max),
    random_float(min, max),
  ];
}

/** @returns {Vec3F} */
function vec3_sub(v1, v2) {
  return [v1[0] - v2[0], v1[1] - v2[1], v1[2] - v2[2]];
}

/** @returns {Vec3F} */
function vec3_add(v1, v2) {
  return [v1[0] + v2[0], v1[1] + v2[1], v1[2] + v2[2]];
}

function vec3_dot(v1, v2) {
  return v1[0] * v2[0] + v1[1] * v2[1] + v1[2] * v2[2];
}

function vec3_normalize(v) {
  return Math.sqrt(vec3_dot(v, v));
}

export function create_world() {
  const origin = [4, 0.2, 0];

  const world = make_geometry_list([]);

  world.push(
    make_geometry({
      kind: "sphere",
      center: [0, -1000, 0],
      radius: 1000.0,
      material: {
        kind: "lambertian",
        albedo: {
          kind: "checker_texture",
          even: {
            kind: "constant_texture",
            color: [0.2, 0.3, 0.1],
          },
          odd: {
            kind: "constant_texture",
            color: [0.9, 0.9, 0.9],
          },
        },
      },
    })
  );

  for (let a = -10; a < 10; a++) {
    for (let b = -10; b < 10; b++) {
      const choose_material = random_float();
      const center = make_vec3f([
        a + 0.9 * random_float(),
        0.2,
        b + 0.9 * random_float(),
      ]);

      if (vec3_normalize(vec3_sub(center, origin)) > 0.9) {
        if (choose_material < 0.8) {
          world.push(
            make_geometry({
              kind: "moving_sphere",
              center0: center,
              center1: vec3_add(center, [0, random_float(0, 0.5), 0]),
              time0: 0.0,
              time1: 1.0,
              radius: 0.2,
              material: {
                kind: "lambertian",
                albedo: {
                  kind: "constant_texture",
                  color: [
                    random_float() * random_float(),
                    random_float() * random_float(),
                    random_float() * random_float(),
                  ],
                },
              },
            })
          );
        } else if (choose_material < 0.95) {
          world.push(
            make_geometry({
              kind: "sphere",
              center,
              radius: 0.2,
              material: {
                kind: "metal",
                albedo: {
                  kind: "constant_texture",
                  color: random_vec3(0.5, 1.5),
                },
                fuzz: 0.5 * random_float(),
              },
            })
          );
        } else {
          world.push(
            make_geometry({
              kind: "sphere",
              center,
              radius: 0.2,
              material: {
                kind: "dielectric",
                ir: 1.5,
              },
            })
          );
        }
      }
    }
  }

  world.push(
    make_geometry({
      kind: "sphere",
      center: [0.0, 1.0, 0.0],
      radius: 1.0,
      material: {
        kind: "dielectric",
        ir: 1.5,
      },
    })
  );

  world.push(
    make_geometry({
      kind: "sphere",
      center: [-4, 1, 0],
      radius: 1,
      material: {
        kind: "lambertian",
        albedo: {
          kind: "constant_texture",
          color: [0.4, 0.2, 0.1],
        },
      },
    })
  );

  world.push(
    make_geometry({
      kind: "sphere",
      center: [4, 1, 0],
      radius: 1,
      material: {
        kind: "metal",
        albedo: {
          kind: "constant_texture",
          color: [0.7, 0.6, 0.5],
        },
        fuzz: 0.0,
      },
    })
  );

  return world;
}

const size = {
  width: 1280,
  height: 720,
  aspect: function () {
    return this.width / this.height;
  },
};

export default make_project({
  name: "scene1",
  settings: {
    output_dir: "./output",
    width: size.width,
    height: size.height,
    nsamples: 20,
    max_depth: 15,
  },
  scene: {
    camera: {
      look_from: [13, 2, 3],
      look_at: [0, 0, 0],
      view_up: [0, 1, 0],
      vertical_fov: 20,
      aspect: size.aspect(),
      aperture: 0.0,
      focus_dist: 10.0,
      time0: 0.0,
      time1: 1.0,
    },
    sky: {
      kind: "solid",
      background: [0.7, 0.8, 1.0],
    },
    world: create_world(),
  },
});
